default/AI/MoveUtilsAI.py

```diff
@@ -165,9 +165,24 @@ def get_nearest_supplied_system(start_system_id):
 
 
 def get_nearest_drydock_system_id(start_system_id):
-    """ Return systemAITarget of nearest supplied system from starting system startSystemID."""
-    drydock_system_ids = ColonisationAI.empire_dry_docks.keys()
+    """Get system_id of nearest drydock capable of repair.
+
+    This function does not consider starlane obstruction!
+
+     :param start_system_id: current location of fleet - used to find closest target
+     :type start_system_id: int
+     :return: closest system_id capable of repairing
+     :rtype: int
+     """
+    # TODO consider starlane obstruction
     universe = fo.getUniverse()
+    drydock_system_ids = set()
+    for sys_id, pids in ColonisationAI.empire_dry_docks.iteritems():
+        for pid in pids:
+            planet = universe.getPlanet(pid)
+            if planet and planet.currentMeterValue(fo.meterType.happiness) >= 5:
```

Comment:
rather than just checking the current happiness meter value here, I would suggest also checking the target happiness along with expected travel time, and also checking that the system is not blockaded.